src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/TbsCertificate.cs

```diff
@@ -0,0 +1,243 @@
+            extensions      [3]  Extensions OPTIONAL
+                                -- If present, version MUST be v3 --  }
+    */
+    internal sealed class TbsCertificate
+    {
+        public byte? Version { get; set; }
+        public byte[] SerialNumber { get; set; }
+        public byte[] SignatureAlgorithm { get; set; }
+        public X500DistinguishedName Issuer { get; set; }
+        public DateTimeOffset NotBefore { get; set; }
+        public DateTimeOffset NotAfter { get; set; }
+        public X500DistinguishedName Subject { get; set; }
+        public PublicKey PublicKey { get; set; }
+        public byte[] IssuerUniqueId { get; set; }
+        public byte[] SubjectUniqueId { get; set; }
+        public Collection<X509Extension> Extensions { get; } = new Collection<X509Extension>();
+
+        private byte[] Encode(X509SignatureGenerator signatureGenerator, HashAlgorithmName hashAlgorithm)
+        {
+            // State validation should be runtime checks if/when this becomes public API
+            Debug.Assert(Subject != null);
+            Debug.Assert(PublicKey != null);
+
+            // Under a public API model we could allow these to be null for a self-signed case.
+            Debug.Assert(SerialNumber != null);
+            Debug.Assert(Issuer != null);
+
+            if (SignatureAlgorithm != null)
+            {
+                ValidateSignatureAlgorithm();
+            }
+
+            List<byte[][]> encodedFields = new List<byte[][]>();
+
+            byte version = GetEffectiveVersion();
+
+            if (version != 0)
+            {
+                byte[][] encodedVersion = DerEncoder.ConstructSegmentedSequence(
+                    DerEncoder.SegmentedEncodeUnsignedInteger(new[] { version }));
+
+                encodedVersion[0][0] = DerSequenceReader.ContextSpecificConstructedTag0;
+
+                encodedFields.Add(encodedVersion);
+            }
+
+            encodedFields.Add(DerEncoder.SegmentedEncodeUnsignedInteger(SerialNumber));
+
+            // SignatureAlgorithm: Use the specified value, or ask the generator (without mutating the class)
+            byte[] signatureAlgorithm = SignatureAlgorithm ?? signatureGenerator.GetSignatureAlgorithmIdentifier(hashAlgorithm);
+            encodedFields.Add(signatureAlgorithm.WrapAsSegmentedForSequence());
+
+            // For public API allowing self-sign ease-of-use, this could be (Issuer ?? Subject).
+            encodedFields.Add(Issuer.RawData.WrapAsSegmentedForSequence());
+
+            encodedFields.Add(
+                DerEncoder.ConstructSegmentedSequence(
+                    EncodeValidityField(NotBefore, nameof(NotBefore)),
+                    EncodeValidityField(NotAfter, nameof(NotAfter))));
+
+            encodedFields.Add(Subject.RawData.WrapAsSegmentedForSequence());
+
+            encodedFields.Add(PublicKey.SegmentedEncodeSubjectPublicKeyInfo());
+
+            if (IssuerUniqueId != null)
```

Comment:
Done.  Sort of.  Since there's no way to set these two fields I've removed them; and v3 is now hardcoded by the caller, but I added an assert in the extensions-if as a note for the future